### PR TITLE
Add platform independent renderer frontend

### LINKED ISSUE #77: Add platform independent renderer frontend
Currently in order to access renderer user has to write code like this:
```rust
let mut context = self.context.borrow_mut();
match context.video_context {
VideoContext::OpenGL { ref window,
                       ref mut renderer,
                       ref frame,
                       ref mut device,
                        .. } => {
    renderer.submit(frame, device);
    window.swap_buffers().unwrap();
}
#[cfg(windows)]
VideoContext::Direct3D {  } => {
    // Direct3D code
},
VideoContext::Null => (),
}
```
It would make sense to add a renderer frontend, that would own `VideoContext` and deal with all this boilerplate internally.

Also platform independent `amethyst_renderer::Fragment` and `amethyst_renderer::Scene` would be required, since they take `gfx::Resource` type parameter, related to #52.

It should simplify implementation of the rendering system.

### PR DESCRIPTION
**Pull request by nchashch**
Changes:
- Add `renderer` and `asset_manger` modules to `amethyst_context`
- Add `Renderer`, `Fragment`, `FragmentImpl`, `Texture`, and `TextureImpl` types to `context::renderer`
- Add `AssetManager` and `FactoryImpl` types to `context::asset_manager`
- Move `Factory` from `VideoContext` to `FactoryImpl`
- Now `Context` owns `context::renderer::Renderer` and `context::asset_manager::AssetManager` instead of `VideoContext`
- Change some doc comments
- Update `examples/window.rs` example
- Add `examples/sphere.rs` example
- Mark all `Direct3D` match branches with `unimplemented!()` macro

`context::renderer::Renderer` allows:
- setting the rendering pipeline
- adding, deleting, and accesing `Fragment`s, `Light`s, and `Camera`s
- adding and deleting `Target`s and `Scene`s
- submitting the rendering job to `amethyst_renderer::Renderer`

`context::asset_manager::AssetManager` allows:
- loading `Texture`s from pixel data
- creating solid color constant `Texture`s
- loading `Mesh`es from vertex data
- creating `Fragment`s from loaded `Texture`s and `Mesh`es

It should close #77, #79 and #52

### CODE REVIEW

File: src/context/src/renderer.rs [hunk 1]
```diff
@@ -0,0 +1,407 @@
...
+            VideoContext::OpenGL {
+                ref mut factory,
+                ..
+            } => {
+                let (buffer, slice) = factory.create_vertex_buffer_with_slice(&data, ());
+
+                let ka = match ka.texture_impl {
+                    TextureImpl::OpenGL { texture } => texture,
+                    #[cfg(windows)]
+                    TextureImpl::Direct3D {  } => return None,
```
Comment (kvark) [inline src/context/src/renderer.rs]: should be `unimplemented!()` I suppose

File: src/context/src/renderer.rs [hunk 2]
```diff
@@ -0,0 +1,407 @@
...
+        let texture_impl = TextureImpl::OpenGL {
+            texture: texture,
+        };
+        Texture {
+            texture_impl: texture_impl,
+        }
+    }
+
+    /// Create a `Texture` from pixel data.
+    pub fn create_texture(&mut self, kind: Kind, data: &[&[<<ColorFormat as Formatted>::Surface as SurfaceTyped>::DataType]]) -> Option<Texture> {
```
Comment (kvark) [inline src/context/src/renderer.rs]: so, just to be clear, this only supports a single format (`ColorFormat`), right?

File: src/context/src/renderer.rs [hunk 3]
```diff
@@ -0,0 +1,407 @@
...
+                                   ref mut renderer,
+                                   ref frame,
+                                   ref mut device,
+                                   .. } => {
+                renderer.submit(frame, device);
+                window.swap_buffers().unwrap();
+            }
+            #[cfg(windows)]
+            VideoContext::Direct3D {  } => {
+                // stub
```
Comment (kvark) [inline src/context/src/renderer.rs]: unimplemented!()

**Comment by kvark** [comment]
I looked over it and liked what I saw so far. The only thing I wish was specified here is the list of the current capabilities/limitations of the renderer, for clarity.

File: src/context/src/renderer.rs (hunk 2 above)
Comment (nchashch) [inline src/context/src/renderer.rs]: Yes, but it probably won't be hard to make it generic.

**Comment by kvark** [comment]
One thing that worries me is the number of matches over GL/D3D/etc.
If they are really the same but just parametrized by different `gfx::Resources`, perhaps we could have a helper like
```rust
self.video_context.with_factory(|fac| { do something with fac });
self.video_context.with_frame(|fr| { do something with frame });
```
Here, the closures accepted are generic over the `R: gfx::Resources`, so we can call them on the appropriate match arm under the hood.

**Comment by kvark** [comment]
Great work! Why did you go with a macro instead of a helper function as I initially suggested? Were there any troubles?

**Comment by nchashch** [comment]
Yes, I tried implementing it using helper functions, but I could not find a way to pass a `VideoContext` field with a concrete type to a generic `with_` function.

So for example this code won't compile because of mismatched types error, since `frame` has concrete `Frame<gfx_device_gl::Resources>` type and a `Frame<R>` type is expected.
```rust
impl Renderer {
    // ...
    fn with_frame<R, F>(&mut self, f: F)
        where R: gfx::Resources,
              F: FnMut(&mut amethyst_renderer::Frame<R>),
    {
        match self.video_context {
            VideoContext::OpenGL {
                ref mut frame,
                ..
            } => {
                f(frame)
            },
            #[cfg(windows)]
            VideoContext::Direct3D {
            } => {
                unimplemented!();
            },
            VideoContext::Null => (),
        }
    }
}
```
Also if it was possible, I am not sure if rust compiler would be able to infer the types for `R` and `F`, so a match over `VideoContext` would be required anyway to pass the proper types to the method.

**Comment by kvark** [comment]
Ah, I see. So what we need is a generic function of `R` to be passed in, but that means higher-order types... It seems.

**Comment by csherratt** [comment]
LGTM

**Pull request merged**